crates/zkevm_circuits/src/bn254/ec_pairing/implementation.rs

```diff
@@ -0,0 +1,461 @@
+        let mut t4 = t3.double(cs);
+        let mut t4 = t4.double(cs);
+        let mut t5 = t4.mul(cs, &mut t2);
+        let mut y_r_2 = r.y.double(cs);
+        let mut t6 = t1.sub(cs, &mut y_r_2);
+
+        // 9. t9 <- t6 * X_Q; 10. t7 <- X_R * t4; 11. X_T <- t6^2 - t5 - 2t7
+        let mut t9 = t6.mul(cs, &mut q.x);
+        let mut t7 = r.x.mul(cs, &mut t4);
+        let mut x_t = t6.square(cs);
+        let mut x_t = x_t.sub(cs, &mut t5);
+        let mut t7_2 = t7.double(cs);
+        let mut x_t = x_t.sub(cs, &mut t7_2);
+
+        // 12. Z_T <- (Z_R + t2)^2 - Z_R^2 - t3;
+        let mut z_t = r.z.add(cs, &mut t2);
+        let mut z_t = z_t.square(cs);
+        let mut z_t = z_t.sub(cs, &mut z_r_square);
+        let mut z_t = z_t.sub(cs, &mut t3);
+
+        // 13. t10 <- Y_Q + Z_T; 14. t8 <- (t7 - X_T)*t6;
+        let mut t10 = q.y.add(cs, &mut z_t);
+        let mut t8 = t7.sub(cs, &mut x_t);
+        let mut t8 = t8.mul(cs, &mut t6);
+
+        // 15. t0 <- 2*Y_R*t5; 16. Y_T <- t8 - t0; 17. t10 <- t10^2 - Y_Q^2 - Z_T^2;
+        let mut t0 = y_r_2.mul(cs, &mut t5);
+        let y_t = t8.sub(cs, &mut t0);
+        let mut t10 = t10.square(cs);
+        let mut t10 = t10.sub(cs, &mut y_q_square);
+        let mut z_t_square = z_t.square(cs);
+        let mut t10 = t10.sub(cs, &mut z_t_square);
+
+        // 18. t9 <- 2*t9 - t10; 19. t10 <- 2*Z_T*y_P;
+        let mut t9 = t9.double(cs);
+        let t9 = t9.sub(cs, &mut t10);
+        let mut t10 = z_t.mul_c0(cs, &mut p.y);
+        let t10 = t10.double(cs);
+
+        // 20. t6 <- -t6; 21. t1 <- 2*t6*x_P;
+        let mut t6 = t6.negated(cs);
+        let mut t1 = t6.mul_c0(cs, &mut p.x);
+        let t1 = t1.double(cs);
+
+        // Result: T = (X_T, Y_T, Z_T); Line function is l0 + l1*w
+        // where l0 = t10; l1 = t1 + t9*v;
+        Self {
+            c0: t10,
+            c3: t1,
+            c4: t9,
+            point: BN256SWProjectivePointTwisted {
+                x: x_t,
+                y: y_t,
+                z: z_t,
+                _marker: std::marker::PhantomData,
+            },
+            _marker: std::marker::PhantomData,
+        }
+    }
+}
+
+/// Struct for the miller loop evaluation for the BN256 curve.
+/// Here, the Miller loop returns the accumulated f value after the loop
+/// without the final exponentiation.
+pub struct MillerLoopEvaluation<F, CS>
+where
+    F: SmallField,
+    CS: ConstraintSystem<F>,
+{
+    accumulated_f: BN256Fq12NNField<F>,
+    _marker: std::marker::PhantomData<CS>,
+}
+
+impl<F, CS> MillerLoopEvaluation<F, CS>
+where
+    F: SmallField,
+    CS: ConstraintSystem<F>,
+{
+    pub fn get_accumulated_f(&self) -> BN256Fq12NNField<F> {
+        self.accumulated_f.clone()
+    }
+
+    /// This function computes the Miller loop for the BN256 curve, using
+    /// _Algorithm 1_ from https://eprint.iacr.org/2010/354.pdf. Frobenius
+    /// map is taken from https://hackmd.io/@Wimet/ry7z1Xj-2.
+    pub fn evaluate(
```

Comment:
Critical bug: You should have used subgroup check during the Miller loop. Here is the article: https://eprint.iacr.org/2022/352.pdf

Reply:
I have had a look at the `go-ethereum` code for ecpairing (which is most representing protocol rules) and haven't found there subgroup checks. Plus,[ EIP-197](https://eips.ethereum.org/EIPS/eip-197) does not mention it. Here is [article](https://ethresear.ch/t/fast-mathbb-g-2-subgroup-check-in-bn254/13974) regarding this. As I understand, it is not done in protocol, particularly due to significant cost of such.

Reply:
You do need to perform a subgroup check for G2 when working with the Bn256 (or BN254) curve. 
The BN256 curve is designed so that both G1 and G2 have a prime order q. However, because G2 is defined over F_p^2, there are elements on the curve that do not necessarily belong to the prime-order subgroup. The reason a subgroup check is required for G2 is that G2 operates over Fp2, where the elliptic curve has more points (the full curve over Fp2 has order p2−1), and only a subset of those points are in the prime-order subgroup. Therefore, G2 can have points outside the correct prime-order subgroup, making the check necessary.
For G1 in the BN256 (BN254) curve, you do not typically need to perform a subgroup check.
The cofactor for the G1 group is 1, meaning the total number of points on the elliptic curve over Fp is exactly q×1. Therefore, there is no risk of having points in a larger group (i.e., no larger cofactor to worry about), and any valid point generated by the group operation already lies in the correct subgroup of prime order q. ( in contrast, there is bls12 curve which have a lot of cofactors so we need subgroup check for G1 and G2)
But the fact that [EIP-197](https://eips.ethereum.org/EIPS/eip-197) does not mention it is suspicious to me so i decide do some research:
Since small subgroup attack is directly breaking DLA for some curves their coffactor could be huge enough to be safe. This article explain how the same curve but with different parameters could change security level https://eprint.iacr.org/2015/247.pdf and the last author of this blog compute for us the bit of security https://blog.bren2010.io/blog/notes-on-bn256
and this is 2^100 what is quite big for use subgroup check.  I checked our curve and we dont have safe parameters but the security is approximately 2^100. So do we need to add subgroup check in our precompiles? ( checked some zk project with the same curve and they did check)
Why do I suggest to use the subgroup check for BN256 during Miller loop? because if you look at the article which i previously sent, subgroup check uses the twisted frobenius map, the same as miller loop. we can just reuse frob map to check if point in the right subgroup.
Subgroup check for BN256 curve is of the form: twisted_frob(Q) = [6*u^2]*Q so shouldn’t be such expensive